Add composite support to js-bmfbp

kenhkan opened the pull request:
This addresses the remaining items of https://github.com/bmfbp/bmfbp/issues/14.

There are two more scenarios that I've identified. See requirement items 11 and 12. I will address those two in an upcoming commit.

kenhkan commented:
@guitarvydas Please review.

kenhkan commented on js-bmfbp/prototype.js:
```diff
@@ -272,91 +368,197 @@ function bmfbp(schematic) {
+    // calls the main function of this Composite.
+    var compositeSend = null;
+    // We need a way to know when to flush the Sinks.
+    var flushSinks = false;
```
@guitarvydas I forgot to point this out. I ended up using a flag called `flushSinks` for each composite to keep track of that we have packets sent through the sinks. I found that I needed it because dispatch only checks whether there is any ready Parts inside the Composite. Is there another way to do this without using a flag?

kenhkan commented:
@guitarvydas 

I renamed "Part" where it should be "Kind" in https://github.com/bmfbp/bmfbp/pull/19/commits/3fb0bc05c1f398f38ad4c26f3c04695c185f1d7e.

I added a COPY Kind so that packets are no longer shared. See https://github.com/bmfbp/bmfbp/pull/19/commits/aaf1bcffb05a4e13b7b155aab3230685e43b446e.

This raised a question for me. What if I want to pass a complex object as a packet? An example would be an object with circular reference. What about file handles, etc? This is entirely a software-only issue that isn't relevant for hardware. Any insight?

guitarvydas commented:
I've just pushed a branch "drawings".

I wonder if we have a communication problem.  Can you complete the drawing "compositeN"?  Something about the ASCII art seems incomplete (at least to me).

guitarvydas commented:
It is not clear to me (A) what part does a(0) belong to? (B) Is there a an overall parent schematic that wires the various composites together?

Even this small test system is too complicated(!) for me to think about.  I am looking at the ASCII diagram to try to understand what is intended.

kenhkan commented:
I'll create the diagrams and put them in https://github.com/bmfbp/bmfbp/pull/24.

kenhkan commented:
I created the drawing here: https://github.com/bmfbp/bmfbp/pull/19/commits/8750087c39b18141f3bd36f5b448725449287e7d

kenhkan commented:
@guitarvydas I forgot to leave my questions:

I replaced the constant boxes with a rounded rectangle. Feel free to replace it with some other shape. I don't think it should be a rectangle because it's not a Kind.

kenhkan commented:
@guitarvydas 

I added https://github.com/bmfbp/bmfbp/pull/19/commits/53b7f4eb5b8aa0c1141578d63916f0fc7aa36aa7 to address requirements 11 and 12 in the source file.

I have a few questions:

1. How do you usually deal with NC? Do you specify the source/sink/wires and put some kind of sentinel value in place of what the source/sink/wire number would be had it been connected? Right now I'm just detecting for `null` here: https://github.com/bmfbp/bmfbp/pull/19/commits/53b7f4eb5b8aa0c1141578d63916f0fc7aa36aa7#diff-63c091a09cba4dc3a7e70065ce56f6c6R509
2. Somewhat related: I found the need to specify the number of output Pins for each Part because a Pin may not be connected and so there would be no information on the Wire for that Pin, but the kernel needs to somehow account for that that Pin does exist. What's your approach to deal with that? Current solution: https://github.com/bmfbp/bmfbp/pull/19/commits/53b7f4eb5b8aa0c1141578d63916f0fc7aa36aa7#diff-63c091a09cba4dc3a7e70065ce56f6c6R188

kenhkan commented:
@guitarvydas 

I forgot to update `compositeP`. Now updated in https://github.com/bmfbp/bmfbp/pull/19/commits/7d6d242f86458f8c1a6f50f496885f46b191a2ec.

I added the example for composite definition in https://github.com/bmfbp/bmfbp/pull/19/commits/8d3c6663d23aab5e9f8c05c2128fe5f34c2429b9.

kenhkan commented:
@guitarvydas 

With the latest commits, the js-bmfbp example now adheres to the spec that we drafted over the weekend.

I also updated the diagrams and have one question. When a composite's sink is connected to one of its child Parts but the corresponding composite's output pin is not connected, there would be an error. We can either catch the error in the kernel or enforce that the composite's output pins always be wired up. I prefer the latter for consistency since we're enforcing that for sources and sinks that are not connected to any Parts. Thoughts?

guitarvydas commented:
I'm not sure that I understand the question fully.  An N/C for an output of a Composite, is not an error, just an N/C.  One cannot know what kind of component (Composite or Leaf) it is from the "outside" (e.g. wiring the Component in a schematic).

kenhkan commented:
@guitarvydas 

It is not supposed to be an error. What I was saying was that the kernel needs to carry a packet from (1) a child Part, to (2) the sink of the composite, to (3) the output pin of the composite, to (4) the input pin of a Part outside of the composite.

Let's say 1 is connected to 2, but 3 is not connected (i.e. there is no 4). Either the kernel needs to know that 3 is not connected, or that schematically, we enforce that 3 is attached to a wire that is not attached to anything.

The latter case is more consistent with what you have with sources/sinks with dangling wires. I was saying I prefer that on the grounds of consistency. Does that clarify the situation?
